File: 200-400/329. Longest Increasing Path in a Matrix/solution.cpp
class Solution {
private:
    int row[4] = {0,1,0,-1};
    int col[4] = {1,0,-1,0};
    int r,c;
    
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        r = matrix.size(), c = matrix[0].size();
        int dp[r*c];
        memset(dp,-1,sizeof(dp));
        int ans = 0;
        for (int i=0; i<r; i++) {
            for (int j=0; j<c; j++) {
                ans = max(ans,dfs(matrix,i,j,dp));
            }
        }
        return ans;
    }
    
private:
    int dfs(vector<vector<int>>& matrix, int i, int j, int dp[]) {
        if (dp[i*c+j]>-1) return dp[i*c+j];
        int val = matrix[i][j], ans = 0;
        for (int k=0; k<4; k++) {
            int m = i+row[k], n = j+col[k];
            if (m<0 || n<0 || m>=r || n>=c || matrix[m][n]<=val) 
                continue;
            ans = max(ans,dfs(matrix,m,n,dp));
        }
        return dp[i*c+j] = ans+1;
    }
    
};

File: 200-400/329. Longest Increasing Path in a Matrix/solution.js
var longestIncreasingPath = function(matrix) {
    let LIP = new Array(matrix.length)
        .fill()
        .map(() => new Array(matrix[0].length).fill(0)) 

    let max = 0
    for (let i=0; i<matrix.length; i++) {
        for (let j=0; j<matrix[i].length; j++) {
            if(LIP[i][j] === 0) max= Math.max(dfs(matrix,i,j,LIP),max)
        }
    }
    return max
};

let dfs = (grid,i,j,LIP) => {
    if (LIP[i][j] !== 0) return LIP[i][j]

    let max = 1
    let actionSpace = [[i+1,j],[i-1,j],[i,j+1],[i,j-1]]
    for (const [m,n] of actionSpace) {
        if (isValid(grid,m,n) && grid[m][n] > grid[i][j]) {
            max = Math.max(dfs(grid,m,n,LIP)+1,max)
        }
    }
    LIP[i][j] = max
    return max
}

let isValid = (grid,i,j) => {
    return i>=0 &&
        i<grid.length &&
        j>=0 &&
        j<grid[i].length
}
